diff.go

```diff
@@ -1,57 +1,57 @@
 package gosymbol
 
-func (e constant) D(varName VarName) Expr {
+func (e constant) D(v variable) Expr {
 	return Const(0.0)
 }
 
-func (e variable) D(varName VarName) Expr {
-	if varName == e.Name {
+func (e variable) D(v variable) Expr {
+	if v == e {
 		return Const(1.0)
 	} else {
 		return Const(0.0)
 	}
 }
 
-func (e add) D(varName VarName) Expr {
+func (e add) D(v variable) Expr {
 	differentiatedOps := make([]Expr, len(e.Operands))
 	for ix, op := range e.Operands {
-		differentiatedOps[ix] = op.D(varName)
+		differentiatedOps[ix] = op.D(v)
 	}
 	return Add(differentiatedOps...)
 }
 
 // Product rule: D(fghijk...) = D(f)ghijk... + fD(g)hijk... + ....
-func (e mul) D(varName VarName) Expr {
+func (e mul) D(v variable) Expr {
 	terms := make([]Expr, len(e.Operands))
 	for ix := 0; ix < len(e.Operands); ix++ {
-		var productOperands []Expr
+		productOperands := make([]Expr, len(e.Operands))
```

Comment:
I guess this is the bug you were talking about? How did it show itself?

Reply:
there was a warning in vs code, I checked it and indeed it failed. Basically copy only copies up to the minimum len of the two slices. In this case the copy was copying 0 elements because productOperand was a nil slice